### PR TITLE
TTB improvements

### PR DESCRIPTION
**Pull request by amandalund**
These changes improve the accuracy of the thick-target bremsstrahlung approximation in OpenMC. Previously, stopping powers used in the TTB model were pulled from the NIST ESTAR database, which generates data for each element using the density at standard temperature and pressure. However, because the collision stopping power depends on material quantities such as the density effect correction, it should be solved for the material rather than by summing data for the constituent elements. Here, instead of using the ESTAR stopping powers we now calculate the density effect correction and the collision stopping power for each material.

Right now the new data needed is packaged in an HDF5 file to be distributed with OpenMC. It consists of the mean excitation energy taken from ESTAR and electron subshell data (binding energies, number of electrons per subshell, and effective number of conduction electrons) pulled from the MCNP el03 library.

As an example of how these changes can affect photon calculations, the energy spectra for MCNP (red) and OpenMC (blue) are shown below with the relative error (green) plotted alongside the uncertainties. These were generated using the photon physics [validation](https://github.com/openmc-dev/validation) script for a 100 MeV source in hydrogen. With these improvements (top), we see better agreement with MCNP than we did previously (bottom).

h-new
h-old

### CODE REVIEW

**Comment by amandalund** [comment]
@paulromano this means we'll need to generate the photon libraries one more time :)

File: src/material.cpp [hunk 1]
```diff
@@ -481,16 +572,11 @@ void Material::init_bremsstrahlung()
     double Z_eq_sq = 0.0;
     double sum_density = 0.0;
 
-    // Calculate the molecular DCS and the molecular total stopping power using
+    // Get the collision stopping power of the material
+    this->collision_stopping_power(&stopping_power_collision(0), positron);
```
Comment (paulromano) [inline src/material.cpp]: `stopping_power_collision.data()`

File: src/material.cpp [hunk 2]
```diff
@@ -867,6 +948,124 @@ void Material::to_hdf5(hid_t group) const
 // Non-method functions
 //==============================================================================
 
+double sternheimer_adjustment(std::vector<double>& f, std::vector<double>&
```
Comment (paulromano) [inline src/material.cpp]: Mark the two vectors as `const`

File: src/material.cpp [hunk 3]
```diff
@@ -867,6 +948,124 @@ void Material::to_hdf5(hid_t group) const
...
+  }
+  // Did not converge
+  if (iter >= max_iter) {
+    warning("Maximum Newton-Raphson iterations exceeded.");
+    rho = 1.0e-6;
+  }
+  return rho;
+}
+
+double density_effect(std::vector<double>& f, std::vector<double>& e_b_sq,
```
Comment (paulromano) [inline src/material.cpp]: Mark the vectors as `const`

File: src/physics.cpp [hunk 4]
```diff
@@ -265,15 +265,17 @@ void sample_photon_reaction(Particle* p)
...
+    double E_electron = (alpha - alpha_out)*MASS_ELECTRON_EV - e_b;
     int electron = static_cast<int>(ParticleType::electron);
-    p->create_secondary(uvw, E_electron, electron, true);
+    if (E_electron >= settings::energy_cutoff[electron]) { 
+      double mu_electron = (alpha - alpha_out*mu)
+        / std::sqrt(alpha*alpha + alpha_out*alpha_out - 2.0*alpha*alpha_out*mu);
+      double uvw[3];
+      std::copy(p->coord[0].uvw, p->coord[0].uvw + 3, uvw);
+      rotate_angle_c(uvw, mu_electron, &phi);
+      p->create_secondary(uvw, E_electron, electron, true);
```
Comment (paulromano) [inline src/physics.cpp]: This bit will need some refactoring after #1183 is merged:
```C++
Direction u = rotate_angle(p->u(), mu_electron, &phi);
p->create_secondary(u, E_electron, Particle::Type::electron);
```

**Comment by paulromano** [comment]
@amandalund Also, once the PR is approved, I'll switch the data and re-run the tests. Otherwise updating the data might screw with other PRs in progress.

**Comment by amandalund** [comment]
Right -- there are still differences in the TTB and pair production models, so I'm expecting some discrepancy.

**Comment by paulromano** [comment]
@amandalund Just a heads up -- looks like there's a merge conflict in physics.cpp

File: src/physics.cpp [hunk 5]
```diff
@@ -264,12 +264,15 @@ void sample_photon_reaction(Particle* p)
...
 
     // Create Compton electron
-    double E_electron = (alpha - alpha_out)*MASS_ELECTRON_EV - e_b;
-    double mu_electron = (alpha - alpha_out*mu)
-      / std::sqrt(alpha*alpha + alpha_out*alpha_out - 2.0*alpha*alpha_out*mu);
     double phi = 2.0*PI*prn();
-    Direction u = rotate_angle(p->u(), mu_electron, &phi);
-    p->create_secondary(u, E_electron, Particle::Type::electron);
+    double E_electron = (alpha - alpha_out)*MASS_ELECTRON_EV - e_b;
+    int electron = static_cast<int>(ParticleType::electron);
```
Comment (paulromano) [inline src/physics.cpp]: This needs to be `Particle::Type::electron`

File: src/material.cpp [hunk 6]
```diff
@@ -451,6 +451,97 @@ void Material::init_thermal()
...
+
+  // Get the Sternheimer adjustment factor
+  double rho = sternheimer_adjustment(f, e_b_sq, e_p_sq, n_conduction, log_I,
+    1.0e-6, 100);
+
+  // Classical elctron radius in cm
+  double r_e = PLANCK_C / (2.0e8 * PI * FINE_STRUCTURE * MASS_ELECTRON_EV);
+
+  // Constant in expression for collision stopping power
+  double c = 2.0e24 * PI * r_e * r_e * MASS_ELECTRON_EV * electron_density;
```
Comment (paulromano) [inline src/material.cpp]: Why 2.0e24?

File: src/material.cpp [hunk 7]
```diff
@@ -451,6 +451,97 @@ void Material::init_thermal()
...
+  double e_p_sq = PLANCK_C * PLANCK_C * PLANCK_C * N_AVOGADRO *
+    electron_density * density / (2.0 * PI * PI * FINE_STRUCTURE *
+    MASS_ELECTRON_EV * mass_density);
+
+  // Get the Sternheimer adjustment factor
+  double rho = sternheimer_adjustment(f, e_b_sq, e_p_sq, n_conduction, log_I,
+    1.0e-6, 100);
+
+  // Classical elctron radius in cm
+  double r_e = PLANCK_C / (2.0e8 * PI * FINE_STRUCTURE * MASS_ELECTRON_EV);
```
Comment (paulromano) [inline src/material.cpp]: Consider:
```C++
constexpr double CM_PER_ANGSTROM {1.0e-8};
constexpr double r_e = CM_PER_ANGSTROM * PLANCK_C / (2.0 * PI * FINE_STRUCTURE * MASS_ELECTRON_EV)
```
Compilers will do constant folding, so better to be more expressive here.

**Comment by amandalund** [comment]
The photon production regression test uses the broomstick problem, so secondary photons should escape before they have a chance to interact. Though looking at this test again, I notice I have an energy in MeV instead of eV, so I'll update it.

File: docs/source/methods/photon_physics.rst [hunk 8]
```diff
@@ -754,14 +762,97 @@ power is calculated using Bragg's additivity rule as
...
+    :label: density-effect-nubar
+
+    \bar{\nu}_i = \nu_i \rho / \nu_p.
+
+The plasma energy :math:`h\nu_p` of the medium is given by
+
+.. math::
+    :label: plasma-frequency
+
+    h\nu_p = \sqrt{(hc)^2/\pi r_e \rho_m N_A Z / A},
```
[the remark this answers is not in the document]
Comment (amandalund) [reply docs/source/methods/photon_physics.rst]: agreed -- denominator is `pi * A`, I'll make that clear.

File: docs/source/methods/photon_physics.rst [hunk 9]
```diff
@@ -754,14 +762,97 @@ power is calculated using Bragg's additivity rule as
...
+material. In :eq:`density-effect-nubar`, :math:`h\nu_i` is the oscillator
+energy, and :math:`\rho` is an adjustment factor introduced to give agreement
+between the experimental values of the oscillator energies and the mean
+excitation energy. The :math:`l_i` in :eq:`density-effect-correction` are
+defined as
+
+.. math::
+    :label: density-effect-li
+
+    l_i &= (\bar{v}_i^2 + 2/3f_i)^{1/2} ~~~~&\text{for}~~ \bar{v}_i > 0 \\
```
[the remark this answers is not in the document]
Comment (amandalund) [reply docs/source/methods/photon_physics.rst]: yep! thanks

File: src/material.cpp (hunk 6 above)
Comment (amandalund) [reply src/material.cpp]: there's a factor of two and a conversion from cm^2 to barns -- I'll clarify.

**Pull request merged**